asteroid/engine/gan_system.py

```diff
@@ -0,0 +1,172 @@
+from collections import OrderedDict
+import soundfile as sf
+import pytorch_lightning as pl
+import torch
+import os
+from torch_stoi import NegSTOILoss
+from asteroid.losses import PITLossWrapper
+
+
+class GanSystem(pl.LightningModule):
+
+    """ Base class for GAN training systems.
+       Contains a generator a discriminator, two optimizers,
+       a generator loss function, a discriminator loss function,
+       a validation loss function,
+       training and validation dataloaders and learning rate scheduler.
+
+       Args:
+           discriminator (torch.nn.Module): Instance of discriminator (d).
+           generator (torch.nn.Module): Instance of generator (g).
+           opt_d (torch.optim.Optimizer): Instance of optimizer for d.
+           opt_g (torch.optim.Optimizer): Instance of optimizer for g.
+           discriminator_loss (callable): Loss function with signature
+               (est_targets, targets).
+            generator_loss (callable): Loss function with signature
+               (est_targets, targets).
+           train_loader (torch.utils.data.DataLoader): Training dataloader.
+           val_loader (torch.utils.data.DataLoader): Validation dataloader.
+           scheduler_d (torch.optim.lr_scheduler._LRScheduler): Instance
+               of learning rate schedulers for d.
+           scheduler_g (torch.optim.lr_scheduler._LRScheduler): Instance
+               of learning rate schedulers for g.
+           conf: Anything to be saved with the checkpoints during training.
+               The config dictionary to re-instantiate the run for example.
+       """
+
+    def __init__(self, discriminator, generator, opt_d, opt_g,
+                 discriminator_loss, generator_loss, train_loader,
+                 validation_loss=PITLossWrapper(NegSTOILoss(), pit_from='pw_pt'),
+                 val_loader=None, scheduler_d=None, scheduler_g=None,
+                 conf=None):
+
+        super(GanSystem, self).__init__()
+        self.discriminator = discriminator
+        self.generator = generator
+        self.opt_d = opt_d
+        self.opt_g = opt_g
+        self.d_loss = discriminator_loss
+        self.g_loss = generator_loss
+        self.validation_loss = validation_loss
+        self.train_loader = train_loader
+        self.val_loader = val_loader
+        self.scheduler_d = scheduler_d
+        self.scheduler_g = scheduler_g
+        self.conf = conf
+        self.estimates = None
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def training_step(self, batch, batch_nb, optimizer_idx):
+        # Get data from data_loader
+        inputs, targets = batch
+        # Forward inputs
+        self.estimates = self(inputs)
```

Comment:
Why do you need `estimates` as an attribute?

Reply:
Also, I don't know how well this will behave when we train on multi GPU.

Reply:
This was done in case someone wants to use estimates during  on_epoch_ends or on_training_end etc to save examples or see the evolution of one estimates.

Reply:
but wouldn't be better to save only validation examples ?

Reply:
We don't make it mandatory to have a validation step, but we always have a training step.
`GanSystem `could take an arg that specifies if we want to save examples `on_training_end` or `on_epoch_end` but we don't do it in `System `I think I will just remove the attribute.